pyanaconda/modules/subscription/system_purpose.py

```diff
@@ -0,0 +1,134 @@
+#
+# System purpose module.
+#
+# Copyright (C) 2018 Red Hat, Inc.
+#
+# This copyrighted material is made available to anyone wishing to use,
+# modify, copy, or redistribute it subject to the terms and conditions of
+# the GNU General Public License v.2, or (at your option) any later version.
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY expressed or implied, including the implied warranties of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
+# Public License for more details.  You should have received a copy of the
+# GNU General Public License along with this program; if not, write to the
+# Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+# 02110-1301, USA.  Any Red Hat trademarks that are incorporated in the
+# source code or documentation are not subject to the GNU General Public
+# License and may only be used or replicated with the express permission of
+# Red Hat, Inc.
+#
+
+import os
+import json
+
+from pyanaconda.core import util
+from pyanaconda.anaconda_loggers import get_module_logger
+log = get_module_logger(__name__)
+
+VALID_FIELDS_FILE_PATH = "/etc/rhsm/syspurpose/valid_fields.json"
+
+def get_valid_fields(valid_fields_file_path=VALID_FIELDS_FILE_PATH):
+    """Get valid role, sla and usage fields for system purpose use.
+
+    If no valid fields are provided, the fields file is not found or can not be
+    parsed then all three lists will be empty.
+
+    :param str valid_fields_file_path: path to a JSON file holding the valid field listings
+
+    :return: role, sla and usage type lists
+    :rtype: [roles], [slas], [usages]
+    """
+    valid_roles = []
+    valid_slas = []
+    valid_usage_types = []
+    if os.path.exists(valid_fields_file_path):
+        try:
+            with open(valid_fields_file_path, "rt") as f:
+                valid_fields_json = json.load(f)
+                valid_roles = valid_fields_json.get("role", [])
+                valid_slas = valid_fields_json.get("service_level_agreement", [])
+                valid_usage_types = valid_fields_json.get("usage", [])
+        except (IOError, json.JSONDecodeError):
+            log.exception("parsing of syspurpose valid fields file at %s failed", valid_fields_file_path)
```

Comment:
Shouldn't this be re-raised?

Reply:
The system purpose functionality is strictly optional at this stage and the idea is to not crash in case some of it's tooling or data is broken. It will still be readily apparent the list of valid values is broken as the lists in system purpose spoke will be empty & the error will be logged.